### PR TITLE
libc: thread-safe newlib

### PR DESCRIPTION
**Pull request by kaidoho**
The aim of this PR is to utilize the functionalty provided by newlib for thread-safety. 

RFC #21519

### Documentation
[Newlib Documentation on __malloc_lock](https://sourceware.org/newlib/libc.html#g_t_005f_005fmalloc_005flock)

[Newlib Documentation on Reentrancy](https://sourceware.org/newlib/libc.html#Reentrancy)

### Mailing Lists
[Newlib discussion on `--enable-newlib-retargetable-locking`](https://newlib.sourceware.narkive.com/s5jTeAUp/patch-newlib-allow-locking-routine-to-be-retargeted)

[[PATCH, newlib] Allow locking routine to be retargeted](https://sourceware.org/ml/newlib/2016/msg01165.html)

[What are the __retarget_lock functions?](http://sourceware-org.1504.n7.nabble.com/What-are-the-retarget-lock-functions-td531854.html)

### CODE REVIEW

**Comment by kaidoho** [comment]
RFC #21519

File: arch/arm/core/swap.c
Comment (stephanosio) [inline arch/arm/core/swap.c]: Note that `arch_swap` is only used for co-operative task switching. When a task is preempted, this function is not called.

`_impure_ptr` should be updated in `z_arm_pendsv`, which is the actual common task switching function.

Line 228 of `swap_helper.S` would be a good place to add this.

https://github.com/zephyrproject-rtos/zephyr/blob/816d2c47a4d5dad73fa1fab16b3602e1562d4cda/arch/arm/core/swap_helper.S#L225-L229

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: include/kernel.h
Comment (stephanosio) [inline include/kernel.h]: It would be desirable to add a separate newlib config symbol that enables `reent` support (e.g. `CONFIG_NEWLIB_LIBC_REENT`).

This symbol can be `default y if MULTITHREADING` so as to only enable `reent` support when multi-threading is enabled.

File: lib/libc/newlib/libc-hooks.c
Comment (pavlohamov) [inline lib/libc/newlib/libc-hooks.c]: What intent of this change? without weak alias nobody can override syscals

Comment (kaidoho) [reply lib/libc/newlib/libc-hooks.c]: OMG, thanks for pointing me on that. This went errornously within the commit - the `__weak` is back now.

File: include/kernel.h
Comment (kaidoho) [reply include/kernel.h]: If there is definitelly only one thread running I think that's a good idea.

File: lib/libc/newlib/libc-hooks.c
Comment (andyross) [inline lib/libc/newlib/libc-hooks.c]: Same comments here as in the other init function.

File: lib/libc/newlib/libc-hooks.c
Comment (andyross) [inline lib/libc/newlib/libc-hooks.c]: I'm not understanding this function.  You're taking a lock pointer as an argument (where is that defined?), but then throwing that value away and replacing it with a new heap block (which may be null, and is unchecked), then dereferencing whatever pointer happened to be stored in that uninitialized heap block to store a pointer into the same block?

I'm guessing that what you really want to be doing is allocating a block containing just the sem/mutex union and assigning that through the opaque pointer you're being passed?

Why is there a header containing pSemOrMtx if the pointer always points to the byte after its own address?  Why not just cast the struct address in the first place?

**Review by andyross — CHANGES_REQUESTED** [review]
The feature seems sound (though I'm no expert on the newlib locking design), but the memory storage for the locks seems kinda wrong?

File: lib/libc/newlib/libc-hooks.c
Comment (kaidoho) [reply lib/libc/newlib/libc-hooks.c]: Newlib - not my idea

File: lib/libc/newlib/libc-hooks.c
Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c]: we do have a pool of objects here, slab would be better...however, mem slabs can't be used from user mode. sys_mem_pool can, the whole object lives in user memory (we route it to the libc memory domain with NEWLIB_LOCK_SECTION)

File: lib/libc/newlib/libc-hooks.c
Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c]: > You're taking a lock pointer as an argument (where is that defined?), but then throwing that value away and replacing it with a new heap block

Yeah this is confusing me too, some more detail on the intention here would be helpful, maybe leave a comment if this is truly correct (although right now it looks like the allocated lock simply leaks)

**Review by tyrantlizard77 — COMMENTED** [review]
thanks for looking into this, our newlib bindings have needed some attention for a while.

File: include/kernel.h
Comment (andrewboie) [reply include/kernel.h]: yeah this would be good, we have some use-cases for disabling mulithreading and using Zephyr more like a HAL (bootloaders, for example)

File: arch/arm/core/swap.c
Comment (andrewboie) [reply arch/arm/core/swap.c]: do we need to make this change in the context switch code for all arches or is there something special about ARM that _impure_ptr needs to be updated like this?

Comment (stephanosio) [reply arch/arm/core/swap.c]: > do we need to make this change in the context switch code for all arches

Yes, `struct reent` is basically per-thread context for newlib.

Comment (stephanosio) [reply arch/arm/core/swap.c]: > also what about SMP?

For ARM, SMP is not supported at the moment. I will look into this in the future.

File: lib/libc/newlib/libc-hooks.c
Comment (andrewboie) [inline lib/libc/newlib/libc-hooks.c]: what happens if this block isn't compiled? (i.e. someone set the dynamic lock mem size to 0)

File: arch/arm/core/swap.c
Comment (andrewboie) [reply arch/arm/core/swap.c]: these changes impact all arches so I think we will need to close on this before it can be merged.

I don't see how a global _impure_ptr updated on context switch could ever work in an SMP system, surely newlib has something for this...does this need to be stored in threa-local storage?

Comment (andrewboie) [reply arch/arm/core/swap.c]: perhaps we need to override __getreent()?

Comment (stephanosio) [reply arch/arm/core/swap.c]: > perhaps we need to override __getreent()?

Looks like that would be the correct approach for SMP.

https://github.com/eblot/newlib/blob/2a63fa0fd26ffb6603f69d9e369e944fe449c246/newlib/libc/sys/linux/linuxthreads/getreent.c#L5-L10

One problem I see is that `__getreent` is not declared `__weak`, so it would not be override-able.

https://github.com/eblot/newlib/blob/2a63fa0fd26ffb6603f69d9e369e944fe449c246/newlib/libc/reent/getreent.c#L10-L14

I wonder Zephyr should create a separate fork of newlib for this.

File: include/kernel.h
Comment (kaidoho) [reply include/kernel.h]: How do you think about about the other switch to reserve memory for the locks? Is it ok, to have one? Would you set the default value to 0? Actually, I dont like to set it to 0 as only the malloc_lock hooks work in this szenario, but I was concerned that existing applications could run out of memory in case I choose a too little default value

**Comment by kaidoho** [comment]
@stephanosio @andyross @andrewboie 

> do we need to make this change in the context switch code for all arches or is there something special about ARM that _impure_ptr needs to be updated like this?
> 
> also what about SMP?

This has an impact on **all** architectures. As described within the RFC, I've only added ARM (checkpatch shall complain until all others are there too, to avoid having a partial implementation going into Zephyr).

**My main intent was to bring up the issue,** perhaps it wasn't a good idea to support the RFC with a PR as it draws more attention to the implementation than the feature. 

Let's get one step back:

- Do you agree that it would be great to have a thread-safe newlib?
- If so, do you want me to work on this PR, or do you rather prefer to work on it yourself on your own fork/branch? I don't mind working on it - got a few days off and want to have this feature, but I think it is way more efficient if you guys take over and make something neat (instead of having to point some novice to the coding guidelines) - no worries to drop this PR in favour of something better.

Now that I have your attention ;), @stephanosio mentioned to patch newlib. That's partially what I considered when mentioning the alternative RTEMS route to go. Having a OS dependend toolchain and droping the support for GNU ARM embedded. Why dropping the support? Well, the next issue that will arise is within libstdc++:

> The C++ library string functionality requires a couple of atomic operations to provide thread-safety. If you don't take any special action, the library will use stub versions of these functions that are not thread-safe. They will work fine, unless your applications are multi-threaded.

> If you want to provide custom, safe, versions of these functions, there are two distinct approaches. One is to provide a version for your CPU, using assembly language constructs. The other is to use  the thread-safety primitives in your operating system.

[https://gcc.gnu.org/onlinedocs/libstdc++/manual/internals.html#internals.thread_safety](https://gcc.gnu.org/onlinedocs/libstdc++/manual/internals.html#internals.thread_safety)

`No strings == no go` to me.

Those functions go **in** libstdc++ there are no simple hooks - one has to either add a full OS / thread model to GCC / libstdc++, or tweak the single thread implementation by adding hooks which we can use like the newlib stuff.

**Comment by stephanosio** [comment]
> Do you agree that it would be great to have a thread-safe newlib?

Not just great, it is absolutely imperative if we are going to do anything useful with the newlib.

Maybe #21519 should be labeled a "bug" and "high priority" since this issue practically renders the newlib useless?

I can see that there are many projects that require the newlib (e.g. net and gui) and this means that there is the possibility of them "randomly" crashing from the thread safety issues at any moment.

> Those functions go **in** libstdc++ there are no simple hooks - one has to either add a full OS / thread model to GCC / libstdc++, or tweak the single thread implementation by adding hooks which we can use like the newlib stuff.

@pabigot This sounds like something that must be addressed before we can say C++ is supported in the Zephyr, alongside many other issues.

**Comment by pabigot** [comment]
> @pabigot This sounds like something that must be addressed before we can say C++ is supported in the Zephyr, alongside many other issues.

Agreed, added to #18554.

Zephyr has some features that make it difficult to guarantee mutex/thread-safety, regardless of language: ZLIs and meta-IRQs.  For the purposes of newlib support we can ignore them.

File: lib/libc/newlib/libc-hooks.c
Comment (kaidoho) [reply lib/libc/newlib/libc-hooks.c]: I used malloc in first place, but it's kind of heavy. Is there any action for me?

File: lib/libc/newlib/libc-hooks.c
Comment (kaidoho) [reply lib/libc/newlib/libc-hooks.c]: Changed that

File: lib/libc/newlib/libc-hooks.c
Comment (kaidoho) [reply lib/libc/newlib/libc-hooks.c]: > How do you think about about the other switch to reserve memory for the locks? Is it ok, to have one? Would you set the default value to 0? Actually, I dont like to set it to 0 as only the malloc_lock hooks work in this szenario, but I was concerned that existing applications could run out of memory in case I choose a too little default value

File: lib/libc/newlib/libc-hooks.c
Comment (kaidoho) [reply lib/libc/newlib/libc-hooks.c]: Fixed the possible null pointer dereference

> I'm guessing that what you really want to be doing is allocating a block containing just the sem/mutex union and assigning that through the opaque pointer you're being passed?

> Why is there a header containing pSemOrMtx if the pointer always points to the byte after its own address? Why not just cast the struct address in the first place?

Your right, i've changed that.

**Review by pabigot — COMMENTED** [review]
I think this is going to be a good step forward; thanks for taking it on.  Just a couple minor non-blocking comments in addition to ones already raised.

IMO the commit messages don't benefit from (1/2) and (2/2) in the subject line.

It might be worth a link in the commit message to newlib documentation on how to support reentrancy, or at least a reference to the #21519 where there are pointers to such documentation.

In particular while looking at this I wanted to know why there were public symbols being defined with non-Zephyr implementation-reserved identifiers like `__lock___foo`, and had to grep through the newlib source to get an answer.  A comment above the definition noting that these are referenced from newlib when it's built for thread-support (assuming that's true) would help future maintainers understand what these things are.

For style the number of blank lines between definitions isn't consistent (one separation has four).  Also I don't think Zephyr generally adds a space in `(t)x` casts.  Using [uncrustify](https://docs.zephyrproject.org/latest/contribute/index.html#uncrustify) could help reveal issues.

**Comment by kaidoho** [comment]
@pabigot Documentation (at least what exists and is not in source), even the newlib discussions are linked in the RFC - and the first thing I did, was to interlink RFC and PR. I've now added a textual link. Shall I simply copy over the RFC? Yes we require a better documentation as soon as it is clear which way to pursue.  Regarding the libstdc++ which came up during the discussion here - shouldn't we open a separate issue for that?

**Comment by pabigot** [comment]
> @pabigot Documentation (at least what exists and is not in source), even the newlib discussions are linked in the RFC - and the first thing I did, was to interlink RFC and PR. I've now added a textual link and copied over the links to the documentation.

It's nice to have it in the github issue, but when we come back to this in six months for maintenance it'd be nice to have something in the code or commit message.  Going back to the issue and PR given a commit SHA1 is not particularly difficult, but it's not trivial either.  Mentioning the RFC issue number as `#21519` in the commit message would make it easier; a short sentence in the code explaining things might even make it unnecessary.

> Regarding the libstdc++ issue which came up during the discussion here - shouldn't we open a separate issue for that?

Perhaps.  There is a tie to this issue in #18554.  I'm not clear on exactly what else needs to be done for C++ support.  I don't believe we're ever going to get C++ threads to be supported by Zephyr: there's too much resistance to C++ at the project level, and I don't believe the thread model is compatible.

**Comment by carlescufi** [comment]
@kaidoho this PR seems a bit stuck. An option to move it forward is to add the "Dev-review" label for it to be discussed in the dev review meeting. Another one is to continue the discussion here with @stephanosio and @andyross

**Comment by kaidoho** [comment]
@carlescufi you are right, on the one hand I am waiting for directions and on the other I began to look into GCC / newlib to find out what it takes to have them support a Zephyr thread model. This will end in an RFC with the aim to have a custom toolchain for Zephyr. Perhaps, one would implement this PR differently when GCC / newlib has to be patched anyway. So, I think it is best to write the RFC regarding the toolchain, link this RFC/PR, and then see which direction the discussion takes. Ok, or would you do differently?

**Comment by alexanderwachter** [comment]
@stephanosio @andyross @andrewboie @carlescufi @pabigot.
It seems that this PR is still relevant for newlib support. How do we proceed with that?

**Comment by andrewboie** [comment]
> @stephanosio @andyross @andrewboie @carlescufi @pabigot.
> It seems that this PR is still relevant for newlib support. How do we proceed with that?

The issue tracking this is currently tracked as "enhancement" and not "bug", and there isn't pressure applied at release time to resolve it since it doesn't contribute to the release bug count requirements. We rely solely on the motivation of the reporter/author to move it along.

I think it should be promoted to "bug", but agree to scope it for 2.5 and find a dedicate owner to see it through if @kaidoho isn't working on it.

**Pull request closed without merging**
